File: tests/core/contractors/codac2_tests_CtcPolytope.cpp
/** 
 *  Codac tests
 * ----------------------------------------------------------------------------
 *  \date       2024
 *  \author     Simon Rohou
 *  \copyright  Copyright 2024 Codac Team
 *  \license    GNU Lesser General Public License (LGPL)
 */

#include <catch2/catch_test_macros.hpp>
#include <codac2_CtcPolytope.h>
#include <codac2_pave.h>
#include <codac2_Figure3D.h>

using namespace std;
using namespace codac2;

TEST_CASE("CtcPolytope")
{
   std::vector<std::pair<Row,double>> facets 
       { { {1,1,1}, 3.0 },
         { {-1,1,1}, 3.0 },
         { {1,-1,1}, 3.0 },
         { {1,1,-1}, 3.0 },
         { {-1,-1,1}, 3.0 },
         { {-1,1,-1}, 3.0 },
         { {1,-1,-1}, 3.0 },
         { {-1,-1,-1}, 3.0 } };
  Polytope p({{-2,2.0},{-2.0,2.0},{-2.0,2.0}}, facets);
  CtcPolytope c(p);

  IntervalVector x(3);
  c.contract(x);
  CHECK(x == IntervalVector({{-2,2},{-2,2},{-2,2}}));

  x = IntervalVector({{1.02,2.0},{1.02,2.0},{1.02,2.0}}); // possible bug
  c.contract(x);
  CHECK(x.is_empty());

//  auto p_ctc  = pave({{-3,3},{-3,3},{-3,3}}, c, 0.2);
//  Figure3D fig_ctc("Paving contractor");
//  fig_ctc.draw_paving(p_ctc);
}

File: tests/core/contractors/codac2_tests_CtcPolytope.py
#!/usr/bin/env python

#  Codac tests
# ----------------------------------------------------------------------------
#  \date       2025
#  \author     Simon Rohou
#  \copyright  Copyright 2024 Codac Team
#  \license    GNU Lesser General Public License (LGPL)

import unittest
from codac import *

class TestSepPolytope(unittest.TestCase):

  def test_SepPolytope(self):

    # Polytope, defined as a set of facets
    facets =[ [ Row([1,1,1]), 3.0 ],
              [ Row([-1,1,1]), 3.0 ],
              [ Row([1,-1,1]), 3.0 ],
              [ Row([1,1,-1]), 3.0 ],
              [ Row([-1,-1,1]), 3.0 ],
              [ Row([-1,1,-1]), 3.0 ],
              [ Row([1,-1,-1]), 3.0 ],
              [ Row([-1,-1,-1]), 3.0 ] ]

    p = Polytope(IntervalVector([[-2,2],[-2,2],[-2,2]]),facets,True)
    c = CtcPolytope(p)

    x = IntervalVector(3)
    c.contract(x)
    self.assertTrue(x == IntervalVector([[-2,2],[-2,2],[-2,2]]))

    x = IntervalVector([[1.02,2.0],[1.02,2.0],[1.0,2.0]])
    c.contract(x)
    self.assertTrue(x.is_empty())

if __name__ ==  '__main__':
  unittest.main()
